edgelet/edgelet-docker/src/image_prune_data.rs

```diff
@@ -0,0 +1,635 @@
+fn get_images_with_timestamp(
+    image_use_filepath: String,
+) -> Result<HashMap<String, Duration>, Error> {
+    if !std::path::Path::new(&image_use_filepath).exists() {
+        log::info!(
+            "Image garbage collection data file not found; creating file at: {}",
+            image_use_filepath.as_str()
+        );
+        let _file = fs::File::create(image_use_filepath.clone()).map_err(Error::CreateFile)?;
+    }
+
+    let res = fs::read_to_string(image_use_filepath);
+    if let Err(e) = res {
+        let msg = format!("Could not read image persistence data: {}", e);
+        log::error!("{msg}");
+        return Err(Error::FileOperation(msg));
+    }
+
+    let contents = res.expect("Reading image persistence data failed");
+
+    let mut image_map: HashMap<String, Duration> = HashMap::new(); // all image pruning data
+
+    // TL;DR: this dumps pruning data into the image_map, where
+    // Key: Image hash, Value: Timestamp when image was last used (in epoch)
+    let contents = contents
+        .lines()
+        .map(|line| line.split(' ').collect::<Vec<&str>>())
+        .map(|vec| (vec[0].to_string(), vec[1]));
+    for (k, v) in contents {
+        image_map.insert(
+            k,
+            Duration::from_secs(v.parse::<u64>().map_err(Error::ParseIntError)?),
+        );
+    }
+
+    Ok(image_map)
+}
+
+fn write_images_with_timestamp(
+    state_to_persist: &HashMap<String, Duration>,
+    temp_file: String,
+    image_use_filepath: String,
+) -> Result<(), Error> {
+    // write to a temp file and then rename/overwrite to image persistence file (to prevent file write failures or corruption)
+    let mut file = std::fs::File::create(temp_file.clone())
+        .expect("Could not create temporary persistence file");
+
+    for (key, value) in state_to_persist {
+        let image_details = format!("{} {}\n", key, value.as_secs());
+        let res = write!(file, "{}", image_details);
+        if res.is_err() {
+            let msg = format!(
+                "Could not write image:{} with timestamp:{} to store",
+                key,
+                value.as_secs()
+            );
+            return Err(Error::FileOperation(msg));
+        }
+    }
+
+    match fs::rename(temp_file, image_use_filepath) {
+        Ok(_) => {}
+        Err(err) => {
+            return Err(Error::FileOperation(format!(
+                "Could not update garbage collection data {}",
+                err
+            )))
+        }
+    };
+
+    Ok(())
+}
+
+// This method separates out the images to be deleted from the images not to be deleted,
+// and return those as a tuple.
+// It takes as input all the images present on the device (that we know about through an
+// iotedge deployment) and the images currently in-use, along with the minimum "age" for
+//  which the images can stay unused. Any images older than this minim age are marked
+//  for deletion.
+type HashMapTuple = (HashMap<String, Duration>, HashMap<String, Duration>);
```

Comment:
You can just use the tuple directly instead of creating a separate type.

Reply:
Trust me, this was not my idea. **cargo clippy** wouldn't let me. Apparently, it's too complex for Rust...

Reply:
Just add an allow directive for clippy directives that don't make sense.

Reply:
done.